docs/rfcs/016-connection-routing.md

```diff
@@ -0,0 +1,150 @@
+documentation, allowing you to run the examples very quickly. Similar
+to sqlfiddle.com.
+
+PostgreSQL authentication works the same as always. All the different
+PostgreSQL authentication options like SCRAM, kerberos, etc. are
+available. [1]
+
+The third option is to authenticate with github single sign-on. When
+you open the connection in psql, you get a link that you open with
+your browser. Opening the link redirects you to github authentication,
+and lets the connection to proceed. This is also known as "Quick
+connect" [2].
+
+
+## Routing the connection
+
+When a client starts a connection, it needs to be routed to the
+correct PostgreSQL instance. Routing can be done by the proxy, acting
+as a man-in-the-middle, or the connection can be routed at the network
+level based on the hostname or IP address.
+
+Either way, Zenith needs to identify which PostgreSQL instance the
+connection should be routed to. If the instance is not already
+running, it needs to be started. Some connections always require a new
+PostgreSQL instance to be created, e.g. if you want to run a one-off
+query against a particular point-in-time.
+
+The PostgreSQL instance is identified by:
+- zenith account (possibly anonymous)
+- cluster (known as tenant in the storage?)
+- branch or snapshot name
+- timestamp (PITR)
+- primary or read-replica
+- one-off read replica
+- one-off writeable branch
+
+When you are using regular PostgreSQL authentication or anonymous
+access, the connection URL needs to contain all the information needed
+for the routing. With github single sign-on, the browser is involved
+and some details - the zenith account in particular - can be deduced
+from the authentication exchange.
+
+There are three methods for identifying the PostgreSQL instance:
+
+- Browser interaction (quick connect)
+- Options in the connection URL and the domain name
+- A pre-defined endpoint, identified by domain name or IP address
+
+### Quick connect
+
+    postgres://<username>@start.zenith.tech/<dbname>
+
+This gives you a link that you open in browser. Clicking the link
+performs github authentication, and the zenith account name is
+provided to the proxy behind the scenes. The proxy routes the
+connection to the primary PostgreSQL instance in cluster called
+"main", branch "main".
+
+Further ideas:
+- You could pre-define a different target for quick connect
+  connections in the UI.
+- You could have a drop-down in the browser, allowing you to connect
+  to any cluster you want. Quick connect can be like Teleport.
+
+### Connection URL
+
+The connection URL looks like this:
+
+    postgres://<username>@<cluster-id>.db.zenith.tech/<dbname>
+
+By default, this connects you to the primary PostgreSQL instance
+running on the "main" branch in the named cluster [3]. However, you can
+change that by specifying options in the connection URL:
+
+- branch
+- timestamp or LSN
+- snapshot name
+- read replica
+- new read replica dedicated for the connection
```

Comment:
It would be great to assign every option a canonical name right away. We might also want to prefix them.

Reply:
> It would be great to assign every option a canonical name right away.

Done.

> We might also want to prefix them.

This is a bit hacky. The URL has to be of the form:

postgres://alice@cluster-1234.db.zenith.tech/postgres?options=<something>

It's unfortunate that we need "options=" there, but libpq and other clients won't accept arbitrary options directly in the URL. The 'options' is an option that is passed to the server as it is, so we can intercept that in the proxy. All the options that PostgreSQL accepts begin with a dash `-`, because they are nominally command-line options. See https://www.postgresql.org/docs/current/app-postgres.html. Most of the options are actually not allowed to be specified by the client, so the only one that people use in practice is '-c', to set GUCs.

So yeah, perhaps we should prefix these neon-specific options somehow. But that makes them longer, and there isn't really much room for confusion currently. If Postgres changes what options are accepted, that might mix things up, but I don't see that happening any time soon, this hasn't changed for years.